Initial tests

cryogenian opened the pull request:
Smoke tests for file components (There are no tests for search bar) 
Fixed file upload in FireFox. 

What's tested 
* Page loaded
* Redirects to salted hash
* Showing item toolbar on hover and click
* moving down on double click
* mounted database is shown
* File upload 
* File renaming
* File deleting
* Trash folder, hide/show
* Creating new folder 
* Creating new notebook

Currently FireFox and PhantomJS work. 
to run tests in FireFox (if one has FF installed) 
```bash
> npm install 
> bower install
> gulp test-make 
> node test/main.js
```
to run tests in PhantomJS (they will fail to rename item) 
```bash
> npm install 
> bower install 
> npm install -g phantomjs
> gulp test-make
> node test/main.js -b phantomjs 
```

jdegoes commented:
We need to incorporate this into the default build, so that when Travis runs, all these tests are run as well. In addition, ideally we could use Gulp to install missing dependencies (e.g. Firefox, PhantomJS, etc.), or at least list them in package.json so they are installed when the user does `npm install`.

cryogenian commented:
* added test run to `.travis.yml`
* added `npm test`
* updated `purescript-webdriver` dependency
* added `gulp run-tests` task 

Unfortunately I have not found any way to install FF via npm.

jdegoes commented:
@cryogenian Have you experimented with the sauce labs account? It should permit cross-browser tests without installing anything.

cryogenian commented:
No, I've focused only on tests. 
I think that this pr isn't ready for CI, but can be easily integrated to development process.

jdegoes commented:
@cryogenian OK, sounds good. We do need to get this running in CI. Well, as soon as CI is working again. :cry:

jdegoes commented:
@garyb Can you build and test and let me know if you have any comments before I merge it?

cryogenian commented:
Submitted #355 for CI integration

cryogenian commented:
Forgot to mention. All configurations live in 'test/config.json'

garyb commented:
I'm not sure it's working correctly for me, I just get this:

```
symlinking
restoring test database
database restored
selenium killed
slamengine killed
```

Do I need to install some other 3rd party thing for Selenium to work properly?

cryogenian commented:
@garyb I think you should modify `test/config.json` (`slamdataUrl`, `connectionString` and `resotreCmd` especially)

garyb commented:
Those all look fine to me, I have the same setup. I can look in mongo and see the populated `testDb` too.

garyb commented:
Oh, launching `ps` won't work though, that's probably it (Windows).

garyb commented:
Running it via `msysgit` gives a slightly more promising result: it hangs after "database restored".

garyb commented:
Definite progress :smile:
```
symlinking
symlinks created
database dropped
restoring test database
database restored
selenium launched
slamengine launched
running tests


Error: Cannot find module 'Test.Selenium'
    at Function.Module._resolveFilename (module.js:336:15)
    at Function.Module._load (module.js:278:25)
    at Module.require (module.js:365:17)
    at require (module.js:384:17)
    at d:\Users\Gary\Documents\GitHub\slamdata\test\main.js:78:21
    at d:\Users\Gary\Documents\GitHub\slamdata\node_modules\mongodb\lib\mongo_client.js:436:11
    at process._tickCallback (node.js:355:11)
Error: Cannot find module 'Test.Selenium'
```

cryogenian commented:
Huh! I'm trying to run it on my windows machine now. symlinking works only if node runs with administrator privilegies

cryogenian commented:
At least I have `firefox setted as browser for tests`

garyb commented:
Ah thanks, running as administrator was the one. I think it's working now... will confirm in a sec.

garyb commented:
:+1: 
```
symlinking
symlinks created
database dropped
restoring test database
database restored
selenium launched
slamengine launched
running tests


firefox setted as browser for tests


navIcon not found
all elements here, page is loaded

CHECK TEST DATABASE IS MOUNTED
all elements here, page is loaded
test database found

CHECK ITEM TOOLBAR
navIcon not found
all elements here, page is loaded
toolbar is hidden
hovered toolbar is visible
bacground-color has been changed after click

CHECKING URL
navIcon not found
all elements here, page is loaded
hash is correct
search path is correct

CHECKING GO DOWN
navIcon not found
all elements here, page is loaded
all elements here, page is loaded
correct hash after goDown

BREADCRUMBS
navIcon not found
all elements here, page is loaded
correct root breadcrumb

CHECKING GO DOWN
navIcon not found
all elements here, page is loaded
all elements here, page is loaded
correct hash after goDown
breadcrumbs are updated
all elements here, page is loaded

CHECKING URL
navIcon not found
all elements here, page is loaded
hash is correct
search path is correct
Ok, went home after click on root breadcrumb

SORTING CHECK

CHECKING GO DOWN
navIcon not found
all elements here, page is loaded
all elements here, page is loaded
correct hash after goDown
all elements here, page is loaded
OK, sort works

FILE UPLOAD

CHECKING GO DOWN
navIcon not found
all elements here, page is loaded
all elements here, page is loaded
correct hash after goDown
Ok, explore notebook created
navIcon not found
all elements here, page is loaded
new items added after upload

MOVE DELETE

CHECKING GO DOWN
navIcon not found
all elements here, page is loaded
all elements here, page is loaded
correct hash after goDown
successfully renamed
item has been deleted

TRASH CHECK

CHECKING GO DOWN
navIcon not found
all elements here, page is loaded
all elements here, page is loaded
correct hash after goDown
Trash is hidden
Trash is visible
all elements here, page is loaded
Deleted item found

NEW FOLDER CHECK

CHECKING GO DOWN
navIcon not found
all elements here, page is loaded
all elements here, page is loaded
correct hash after goDown
all elements here, page is loaded
ok, hash correct

NEW NOTEBOOK CHECK

CHECKING GO DOWN
navIcon not found
all elements here, page is loaded
all elements here, page is loaded
correct hash after goDown
ok, notebook created
navIcon not found
all elements here, page is loaded
OK, new notebook found in parent directory


cleaning
database dropped
selenium killed
slamengine killed
selenium killed
slamengine killed
```

cryogenian commented:
:fireworks:

cryogenian commented:
@jdegoes Is it ok? Will you merge it?

garyb commented:
Hey! Go be on holiday! :smile:

I'm happy with this @jdegoes if that's what you were waiting for.

cryogenian commented:
@garyb Huh! ok :smiley:
